File: phoenix/test/players/achievement_controller_test.exs
defmodule CardsProjectWeb.Players.AchievementControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "name" => "test",
    "description" => "test",
    "points" => 1,
    "is_hidden" => true,
    "rarity" => "Common"
  }

  describe "GET /api/achievements" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/achievements")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "GET /api/achievements?q=test" do
    test "returns 200 with search results", %{conn: conn} do
      conn = get(conn, "/api/achievements?q=test")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/achievements" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/achievements", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
    test "fails when points_positive violated", %{conn: conn} do
      params = Map.put(@valid_params, "points", 0)
      conn = post(conn, "/api/achievements", params)
      assert conn.status in [400, 422]
    end
  end

  describe "GET /api/achievements/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Players.create_achievement(@valid_params)
      conn = get(conn, "/api/achievements/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/achievements/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Players.create_achievement(@valid_params)
      conn = put(conn, "/api/achievements/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end

end

File: spring/src/test/java/cardsproject/controller/players/AchievementControllerTest.java
package cardsproject.controller.players;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class AchievementControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/achievements"))
            .andExpect(status().isOk());
    }

    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/achievements")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"description\": \"test\", \"points\": 1 }"))
            .andExpect(status().isCreated());
    }

    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/achievements/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }

    @Test
    void delete_returns204or404() throws Exception {
        mockMvc.perform(delete("/api/achievements/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 204 || status == 404 || status == 500 || status == 501;
            });
    }
    @Test
    void create_fails_when_points_positive_violated() throws Exception {
        // Achievement must award at least one point → 400 (Bean Validation)
        mockMvc.perform(post("/api/achievements")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"name\": \"test\", \"description\": \"test\", \"rarity\": \"COMMON\", \"isHidden\": true, \"points\": 0 }"))
            .andExpect(status().isBadRequest());
    }
}
